pkg/river/printer/walker.go

```diff
@@ -0,0 +1,281 @@
+package printer
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/grafana/agent/pkg/river/ast"
+	"github.com/grafana/agent/pkg/river/token"
+)
+
+// A walker walks an AST and sends lexical tokens and formatting information to
+// a printer.
+type walker struct {
+	p *printer
+}
+
+func (w *walker) Walk(node ast.Node) error {
+	switch node := node.(type) {
+	case *ast.File:
+		w.walkFile(node)
+	case ast.Body:
+		w.walkStmts(node)
+	case ast.Stmt:
+		w.walkStmt(node)
+	case ast.Expr:
+		w.walkExpr(node)
+	default:
+		return fmt.Errorf("unsupported node type %T", node)
+	}
+
+	return nil
+}
+
+func (w *walker) walkFile(f *ast.File) {
+	w.p.SetComments(f.Comments)
+	w.walkStmts(f.Body)
+}
+
+func (w *walker) walkStmts(ss []ast.Stmt) {
+	for i, s := range ss {
+		var addedSpacing bool
+
+		// Two blocks should always be separated by a blank line.
+		if _, isBlock := s.(*ast.BlockStmt); i > 0 && isBlock {
+			w.p.Write(wsNewline)
+			addedSpacing = true
+		}
+
+		// A blank line should always be added if there is a blank line in the
+		// source between two statements.
+		if i > 0 && !addedSpacing {
+			var (
+				prevLine = ast.EndPos(ss[i-1]).Position().Line
+				curLine  = ast.StartPos(ss[i-0]).Position().Line
+
+				lineDiff = curLine - prevLine
+			)
+
+			if lineDiff > 1 {
+				w.p.Write(wsFormfeed)
+			}
+		}
+
+		w.walkStmt(s)
+
+		// Statements which cross multiple lines don't belong to the same row run.
+		// Add a formfeed to start a new row run if the node crossed more than one
+		// line, otherwise add the normal newline.
+		if nodeLines(s) > 1 {
+			w.p.Write(wsFormfeed)
+		} else {
+			w.p.Write(wsNewline)
+		}
+	}
+}
+
+func nodeLines(n ast.Node) int {
+	var (
+		startLine = ast.StartPos(n).Position().Line
+		endLine   = ast.EndPos(n).Position().Line
+	)
+
+	return endLine - startLine + 1
+}
+
+func (w *walker) walkStmt(s ast.Stmt) {
+	switch s := s.(type) {
+	case *ast.AttributeStmt:
+		w.walkAttributeStmt(s)
+	case *ast.BlockStmt:
+		w.walkBlockStmt(s)
+	}
+}
+
+func (w *walker) walkAttributeStmt(s *ast.AttributeStmt) {
+	w.p.Write(s.Name.NamePos, s.Name, wsVTab, token.ASSIGN, wsBlank)
+	w.walkExpr(s.Value)
+}
+
+func (w *walker) walkBlockStmt(s *ast.BlockStmt) {
+	joined := strings.Join(s.Name, ".")
+
+	// TODO(rfratto): Should blocks have a oneline format if they're short or
+	// empty? e.g.: `empty_block { attr = 5 }`, `empty_block {}`
+
+	w.p.Write(
+		s.NamePos,
+		&ast.IdentifierExpr{Name: joined, NamePos: s.NamePos},
+	)
+
+	if s.Label != "" {
+		label := fmt.Sprintf("%q", s.Label)
+
+		w.p.Write(
+			wsBlank,
+			&ast.LiteralExpr{Kind: token.STRING, Value: label},
+		)
+	}
+
+	w.p.Write(
+		wsBlank,
+		s.LCurly, token.LCURLY, wsIndent,
```

Comment:
This took a bit to understand, but I assume its patterned off Gos printer?

Reply:
yeah, the Write stuff is based off of go/printer. Which part of it took a while to understand? Maybe I can add more comments explaining why something is happening

Reply:
`s.LCurly, token.LCURLY` the initial look was why are we adding two curlys.

Reply:
Done, every `token.Pos` field in the AST is now suffixed with `Pos` (i.e., `LCurlyPos` instead of `LCurly`).